ext/spl/spl_array.c

```diff
@@ -1790,10 +1791,19 @@ SPL_METHOD(Array, unserialize)
 		intern->ar_flags |= flags & SPL_ARRAY_CLONE_MASK;
 		zval_ptr_dtor(&intern->array);
 		ZVAL_UNDEF(&intern->array);
-		if (!php_var_unserialize(&intern->array, &p, s + buf_len, &var_hash)
-				|| (Z_TYPE(intern->array) != IS_ARRAY && Z_TYPE(intern->array) != IS_OBJECT)) {
+
+		if (!php_var_unserialize(&array, &p, s + buf_len, &var_hash)
+				|| (Z_TYPE(array) != IS_ARRAY && Z_TYPE(array) != IS_OBJECT)) {
 			goto outexcept;
 		}
+
+		if (Z_TYPE(array) == IS_ARRAY ||
+			Z_OBJ_HT(array) == &spl_handler_ArrayObject ||
+			Z_OBJ_HT(array) == &spl_handler_ArrayIterator) {
+			spl_array_set_array(object, intern, &array, 0L, 1);
+		} else {
+			ZVAL_COPY(&intern->array, &array);
+		}
```

Comment:
@nikic thanks
1) I've added additional check and fix for this.
2) yeah, it's good idea to use `spl_array_set_array` for normal objects as well. Just implemented it